### ISSUE TITLE
get_value fails with " 403 Client Error: Forbidden\nMESSAGE: Decrypt option requires administrator access for url: " for encrypted keys

### ISSUE THREAD

**Issue by dpasupathi** [StackStorm/st2#3823]
@kami:
Using data store action_service APIs tried to store and retrieve values in action context.
(similar to https://github.com/StackStorm/st2tests/blob/ddb6c8243fcacc03e12cb8739b81b68402705214/packs/fixtures/actions/pythonactions/datastore_test_action.py#L48)

My simple code snippet looks as follows:
self.action_service.set_value(name=lookup_key, value=passwd, local=False, encrypt=True)

user_kv = self.action_service.get_value(name=lookup_key, local=False, decrypt=decrypt)

When i run the action using "st2 run" with admin privilege  the retrieval failed with following error:
"""
Retrieving value from the datastore (name=switch.1.2.3.6.passwd)\nst2.actions.python.ABCMeta: ERROR    Exception retrieving value from datastore (name=switch.1.2.3.6.passwd): 403 Client Error: Forbidden\nMESSAGE: Decrypt option requires administrator access for url: http://127.0.0.1:9101/v1/keys/switch.1.2.3.6.passwd?scope=system&decrypt=true\nTraceback (most recent call last):\n  File \"/opt/stackstorm/st2/lib/python2.7/site-packages/st2common/services/datastore.py\", line 98, in get_value\n    kvp = client.keys.get_by_id(id=name, params=params)\n
"""

But i run the same via CLI  "st2 key get" it succeeds.
```
user@ubuntu-bwc:/opt/stackstorm/packs/network_essentials/actions$ st2 key get switch.1.2.3.6.passwd --decrypt

+------------------+----------------------------+

| Property         | Value                      |

+------------------+----------------------------+

| name             | switch.1.2.3.6.passwd      |

| value            | password                   |

| secret           | True                       |

| encrypted        | False                      |

| scope            | st2kv.system               |

| expire_timestamp |                            |

+------------------+----------------------------+
```

***Environment:***

```
user@ubuntu-bwc:/opt/stackstorm/packs/network_essentials/actions$ st2 whoami

Currently logged in as admin

user@ubuntu-bwc:/opt/stackstorm/packs/network_essentials/actions$ st2 role list

+--------------------------+--------------+--------+--------------+

| id                       | name         | system | description  |

+--------------------------+--------------+--------+--------------+

| 59f25901ccad682d45c06810 | admin        | True   | admin        |

| 59f25901ccad682d45c06811 | observer     | True   | observer     |

| 59f25901ccad682d45c06812 | system_admin | True   | system_admin |

+--------------------------+--------------+--------+--------------+
user@ubuntu-bwc:/opt/stackstorm/packs/network_essentials/actions$ st2 role-assignment list --user=admin

+--------------------------+--------------+-------+-----------+--------------------+-------------+

| id                       | role         | user  | is_remote | source             | description |

+--------------------------+--------------+-------+-----------+--------------------+-------------+

| 59f25abcccad6845a0e4d72a | system_admin | admin | False     | assignments/admin. |             |

|                          |              |       |           | yaml               |             |

+--------------------------+--------------+-------+-----------+--------------------+-------------+

```

***Note:***
This happens in the st2 environment installed in VM.  In Vagrant st2 environment with st2admin user i am not seeing this issue.

**Comment by Kami** [comment]
> This happens in the st2 environment installed in VM. In Vagrant st2 environment with st2admin user i am not seeing this issue.

Are you running the same st2 version in both environments? What version are you using?

**Comment by Kami** [comment]
@dpasupathi I just had a look and couldn't reproduce the issue with the latest version with RBAC enabled:

```bash
(virtualenv)vagrant@vagrant-ubuntu-trusty-64:/data/stanley$ python ./st2client/st2client/shell.py whoami
Currently logged in as testu

(virtualenv)vagrant@vagrant-ubuntu-trusty-64:/data/stanley$ python ./st2client/st2client/shell.py key set --scope system --encrypt testkey testvalue
+------------------+--------------------------------------------------------------+
| Property         | Value                                                        |
+------------------+--------------------------------------------------------------+
| name             | testkey                                                      |
| value            | 0083E1033AF711D8EB53593E74779DEB8B06305CBFC5E553F2B8D9156399 |
|                  | 7C3F05DB020BF23075DD4F6D6202C74A459931E8D4729EC5713B3F       |
| expire_timestamp |                                                              |
+------------------+--------------------------------------------------------------+

(virtualenv)vagrant@vagrant-ubuntu-trusty-64:/data/stanley$ python ./st2client/st2client/shell.py key get testkey --decrypt
+------------------+--------------+
| Property         | Value        |
+------------------+--------------+
| name             | testkey      |
| value            | testvalue    |
| secret           | True         |
| encrypted        | False        |
| scope            | st2kv.system |
| expire_timestamp |              |
+------------------+--------------+

(virtualenv)vagrant@vagrant-ubuntu-trusty-64:/data/stanley$ python ./st2client/st2client/shell.py run examples.print_config
.
id: 5a0053800640fd2b5334333b
status: succeeded
parameters: None
result: 
  exit_code: 0
  result: null
  stderr: 'st2.actions.python.PrintConfigAction: AUDIT    Creating new Client object.

    st2.actions.python.PrintConfigAction: AUDIT    Creating new Client object.

    st2.actions.python.PrintConfigAction: AUDIT    Retrieving value from the datastore (name=testkey)

    st2.actions.python.PrintConfigAction: AUDIT    Retrieving value from the datastore (name=testkey)

    '
  stdout: '=========

    {u''region'': u''us-east-1''}

    =========

    testvalue

    '
```

User ``testu`` has an admin role and here is a snippet from the action:

```python
value = self.action_service.get_value('testkey', scope='system',
                local=False, decrypt=True)
        print value
```

**Comment by Kami** [comment]
To make it easier to troubleshoot such issues in the future, I'm adding some new functionality which will allow user to easily retrieve information on how user authenticated (and with which user) via the API and via the action and sensor service.

https://github.com/StackStorm/st2/pull/3831

**Comment by dpasupathi** [comment]
@Kami In both cases the st2 version is 2.5

**Comment by Kami** [comment]
I managed to track it down today. It's a bug in v2.5.0 (it was an unintentional regression introduced quite a while ago - https://github.com/StackStorm/st2/pull/3803/files#diff-e9ab53ae60b863f5f7aea0e9257f6a45L87).

Good news is that we already have a fix in master for that as part of this PR -
 https://github.com/StackStorm/st2/pull/3803 (https://github.com/StackStorm/st2/pull/3803#issuecomment-339895677). This PR actually changes and improves some other things, but it also indirectly fixes that.

First I will add a test case to make sure we don't have such regression in the future again and then I will sync up with @StackStorm/team to decide if we should do v2.6.0 or v2.5.1. If we will do v2.5.1 I will need to re-do the change, but change in #3803 was originally designed for v2.6.0 and it's much bigger and contains other changes.

**Comment by cognifloyd** [comment]
Here's an ugly workaround `/opt/stackstorm/rbac/assignments/action_service.yaml`:
```
---
username: "action_service"
roles:
  - "admin"
```
And if a sensor needs encrypted values, then I would do the same for `sensor_service`.

It makes it so that anything done with via the action_service user (used by the action_service when accessing the data store) such as getting encrypted values is effectively always performed with admin rights without regard to the user initiating the action. Beyond that, the rest of RBAC works as expected.

**Comment by cognifloyd** [comment]
Ah. I see the fix was included in 2.5.1 - I was using 2.5.0.

**Comment by Kami** [comment]
Thanks for confirming - will close this issue.

**Comment by Kami** [comment]
Resolved by #3835.
